enhancements/proposed/20190930-addon-host-os-hooks.md

```diff
@@ -0,0 +1,215 @@
+                - ^.*Red Hat.*$
+              command: echo "I am about to be disabled on RedHat or Ubuntu 18"
+            - getHostOSInfoCommand: uname
+              regex: Linux
+              command: echo "I am about to be disabled on Linux"
+        windows:
+          filteredCommands:
+            - getHostOSInfoCommand: systeminfo | findstr /B /C:"OS Version"
+              # Windows 10 or greater
+              regex: ^.*OS Version:[ \t]+([1]+[0-9])\..*$
+              command: Echo "I am about to be disabled"
+      # (optional) Runs just before the addon is disabled
+      postDisable:
+        # Command to run on all unix operating systems. Overrides filteredCommands
+        unix:
+          command: echo "I am disabled on Linux"
+        # Command to run on all windows operating systems. Overrides filteredCommands
+        windows:
+          command: Echo "I am disabled"
+```
+This configuration is loaded along with the other addon configurations (like whether or not it should be enabled/disabled by default)
+When a addon is disabled, or enabled we check whether or not a configuration exists and which hooks are configured. 
+
+If a "preEnable" hook exits with a status code other than 1 then the addon installation fails with an error message 
+printed to the console and logged
+If a "postEnable" hook exits with a status code other than 0 then the addon remains enabled with an error message
+printed to the console and logged 
+If a "preDisable" hook exits with a status code other than 0 then the addon is still disabled with an error message
+printed to the console and logged 
+If a "preDisable" hook exits with a status code other than 0 then the addon is still disabled with an error message
+printed to the console and logged
+
+### Testing Plan 
+Write integration test for an `example` addon which will install an example set of kubernetes configurations and which has all hooks configured.
+The output of each configuration will be tested and should match the expected result based on the Host OS
+
+Include integration test for failed hooks
+Include an integration test for requesting user input. This should be good enough to test that a addon can request to run a hook with 
+sudo permission
+
+## Alternatives Considered
+
+### Background service that monitors addon status
+Write a wrapper program which runs as a background service and continuously monitors which addons are enabled or disabled by running
+minikube commands on the cli
+
+#### Cons of doing this: 
+1. Any action that should be taken as hook by said background service and which also must have sudo permission 
+   must always have sudo permission. This decreases user security since you will basically have to give a program 
+   permission to run as an admin in the backgroundwhich means it can literally do anything it wants to do.
+2. The preEnable and preDisable hooks could not exist which means that depending on what the application is doing
```

Comment:
Do you mean that they might not exist in the config?

Reply:
I mean a background service rather than a hook, could not intercept an enable or disable command so preEnable and preDisable features could not be developed if a background service is used to respond to the user initiated event of enabling or disabling an addon.